### PR TITLE
Fix AppNavigationController navigation bar transparency

### PR DESCRIPTION
**Pull request by SebastianWild**
## Checklist

* [ ] This PR does not change text that hasn't been signed off with the RKI. Have a look at the pinned issue [440](https://github.com/corona-warn-app/cwa-app-ios/issues)
* [ ] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [ ] If this PR comes from a fork, please [Allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
* [ ] Set a speaking title. Format: {task_name} (closes #{issue_number}). For example: Use logger (closes # 41)
* [ ] [Link your Pull Request to an issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) - Pull Requests that are not linked to an issue with you as an assignee will be closed.
* [ ] Create Work In Progress [WIP] pull requests only if you need clarification or an explicit review before you can continue your work item.
* [ ] Make sure that your PR is not introducing _unnecessary_ reformatting (e.g., introduced by on-save hooks in your IDE)

## Description

Fixes a transparency issue with navigation bars handled by `AppNavigationController`.

### Steps to Reproduce

1. Start on the home screen
2. Scroll down until you can see the FAQ cell
3. Observe that the navigation bar is now translucent instead of completely transparent as when the view was scrolled to the top
4. Tap on the FAQ, full-screen Safari will show
5. Tap done
6. Scroll all the way up
7. Observe that the navigation bar did not become transparent
8. Go to the home screen
9. Navigate back to the app & scroll some
10. Observe that in some cases, the navigation bar becomes completely transparent even when scrolled away from the top

IMG_66F9E4DA53B0-1

IMG_AFA216CD3CC7-1

With the fix in this PR, it should no longer happen.

## Alternative Approaches
We might just be able to not invalidate the observer in `viewWillDisappear(_ :)` - this looks like it fixes the issue as well. Since Swift 4 block based observers will clean themselves up as part of their deist process

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/App/AppNavigationController.swift [hunk 1]
```diff
@@ -39,10 +39,21 @@ class AppNavigationController: UINavigationController {
...
 		scrollViewObserver?.invalidate()
 	}
 
+	override func viewDidAppear(_ animated: Bool) {
+		super.viewDidAppear(animated)
+
+		if let topViewController = topViewController {
+			// Since we invalidate the observer on `viewWillDisappear()`, we need to establish it again here.
+			// Else the opacity is not set correctly upon returning to this screen.
+			// Ex. this happens when the FAQ page is presented.
```
Comment (melloskitten) [inline src/xcode/ENA/ENA/Source/Scenes/App/AppNavigationController.swift:52]: thank you so much for adding comments to your pull request. it is GREATLY appreciated. 👏

File: src/xcode/ENA/ENA/Source/Scenes/App/AppNavigationController.swift [hunk 2]
```diff
@@ -103,6 +108,25 @@ extension AppNavigationController: UINavigationControllerDelegate {
...
 		})
 	}
+
+	/// If the passed in `UIViewController` is a `NavigationBarOpacityDelegate` and contains a `UIScrollView`,
+	/// register an observer for the `contentOffset` property so that the navigation bar's `backgroundAlpha` is set as the controller scrolls.
+	///
+	/// - parameter viewController: The controller to regsiter the observer for
+	private func observeScrollView(of viewController: UIViewController) {
+		guard
+			let opacityDelegate = viewController as? NavigationBarOpacityDelegate,
```
Comment (melloskitten) [inline src/xcode/ENA/ENA/Source/Scenes/App/AppNavigationController.swift:118]: this is an interesting way to break the guard statement, but it looks very appealing tbh. Thanks for using this trick, I learnt something new 👍

**Review by melloskitten — APPROVED** [review]
Thank you for your pull request, for me it looks very good. Also, big thanks for adding comments to your code! 💯 happy merging.

**Review by johannesrohwer — APPROVED** [review]
Thanks for your pull request! 🎉  very detailed steps to reproduce the issue and a clean implementation including comments. Great job! 🥇

**Comment by haosap** [comment]
Hi @SebastianWild ,
Thanks. Does it also solve the issue on JIRA1739?

**Comment by SebastianWild** [comment]
> Hi @SebastianWild ,
> Thanks. Does it also solve the issue on JIRA1739?

Yep

**Comment by haosap** [comment]
Hi @SebastianWild ,
Thanks. Please merge it, then.

**Pull request merged**